task/thermal/src/bsp/grapefruit.rs

```diff
@@ -0,0 +1,140 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! BSP for Medusa
+
+use crate::control::{
+    ChannelType, ControllerInitError, Device, Emc2305State, FanControl, Fans,
+    InputChannel, PidConfig, TemperatureSensor,
+};
+use task_sensor_api::SensorId;
+use task_thermal_api::ThermalProperties;
+use userlib::units::Celsius;
+use userlib::TaskId;
+use userlib::UnwrapLite;
+
+include!(concat!(env!("OUT_DIR"), "/i2c_config.rs"));
+use i2c_config::devices;
+use i2c_config::sensors;
+
+////////////////////////////////////////////////////////////////////////////////
+// Constants!
+
+// Air temperature sensors, which aren't used in the control loop
+const NUM_TEMPERATURE_SENSORS: usize = 0;
+
+// Temperature inputs (I2C devices), which are used in the control loop.
+pub const NUM_TEMPERATURE_INPUTS: usize = 1;
+
+// External temperature inputs, which are provided to the task over IPC
+pub const NUM_DYNAMIC_TEMPERATURE_INPUTS: usize = 0;
+
+// Number of individual fans
+pub const NUM_FANS: usize = 5;
+
+// Run the PID loop on startup
+pub const USE_CONTROLLER: bool = true;
+
+////////////////////////////////////////////////////////////////////////////////
+
+bitflags::bitflags! {
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    pub struct PowerBitmask: u32 {
+        const ON = 0b00000001;
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum SeqError {}
+
+#[allow(dead_code)]
+pub(crate) struct Bsp {
+    pub inputs: &'static [InputChannel],
+    pub dynamic_inputs: &'static [SensorId],
+    pub misc_sensors: &'static [TemperatureSensor],
+    pub pid_config: PidConfig,
+
+    fctrl: Emc2305State,
+}
+
+impl Bsp {
+    pub fn fan_control(
+        &mut self,
+        fan: crate::Fan,
+    ) -> Result<crate::control::FanControl<'_>, ControllerInitError> {
+        Ok(FanControl::Emc2305(
+            self.fctrl.try_initialize()?,
+            fan.0.try_into().unwrap_lite(),
+        ))
+    }
+
+    pub fn for_each_fctrl(
+        &mut self,
+        mut fctrl: impl FnMut(FanControl<'_>),
+    ) -> Result<(), ControllerInitError> {
+        match self.fan_control(0.into()) {
+            Ok(c) => {
+                fctrl(c);
+                Ok(())
+            }
+            Err(e) => Err(e),
+        }
```

Comment:
style nit, take it or leave it: isn't this just 
```suggestion
        self.fan_control(0.into()).map(fctrl)
```